jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -78,14 +78,41 @@ class S3TreeDirectoryNode(private val bucket: S3VirtualBucket, parent: S3TreeDir
 
 private val fileTypeRegistry = FileTypeRegistry.getInstance()
 
-class S3TreeObjectNode(bucketName: String, parent: S3TreeDirectoryNode?, key: String, val size: Long, val lastModified: Instant) :
-    S3TreeNode(bucketName, parent, key) {
+open class S3TreeObjectNode(val bucket: S3VirtualBucket, parent: S3TreeDirectoryNode?, key: String, val size: Long, val lastModified: Instant) :
+    S3TreeNode(bucket.name, parent, key) {
 
+    var showHistory: Boolean = false
     private val fileType = fileTypeRegistry.getFileTypeByFileName(name)
 
     init {
         fileType.takeIf { it !is UnknownFileType }?.icon.let { icon = it }
     }
+
+    override fun getChildren(): Array<S3TreeNode> {
+        if (showHistory) {
+            val response = runBlocking {
+                bucket.listVersionObjects(key)
+            }
+            return (
+                response
+                    .versions()
+                    ?.map { S3TreeObjectVersionNode(bucket, parent, key, it.size(), it.lastModified(), it.versionId()) as S3TreeNode }
+                    ?: emptyList()
+                )
+                .toTypedArray()
+        }
+        return emptyArray()
+    }
+}
+
+class S3TreeObjectVersionNode(bucket: S3VirtualBucket, parent: S3TreeDirectoryNode?, key: String, size: Long, lastModified: Instant, val versionId: String) :
+    S3TreeObjectNode(bucket, parent, key, size, lastModified) {
```

Comment:
should `getChildren` be overridden to always be `emptyArray`?

Reply:
yes ,it should be there